### ISSUE TITLE
dep.getResourceIdentifier is not a function (bundle android release)

### ISSUE THREAD

**Issue by FredyC** [callstack/haul#423]
### Current Behavior
The following command is being executed during `:app:bundleReleaseJsAndAssets` and fails with cryptic error `dep.getResourceIdentifier is not a function`. No stacktrace whatsoever. _I've stripped long paths for brevity._
```bash
node node_modules/haul/bin/cli.js bundle \
--platform android --dev false --reset-cache \
--entry-file index.android.js \
--bundle-output android\app\build\intermediates\assets\release\index.android.bundle \
--assets-dest android\app\build\intermediates\res\merged\release
```

### Expected Behavior
Would be lovely to get the bundle :) I do suspect something wrong with entry file seeing command above that there is `index.android.js` instead of `index.ts` I have in my webpack.config.js. This part is somewhat confusing. I even tried changing it to `index.ts`, but no change of the result.

It's also strange why I am getting warning about deprecated configuration style.

### Haul Configuration (webpack.haul.js)
```js
const Dotenv = require('dotenv-webpack')
const path = require('path')
const tsconfig = require('./tsconfig.json')

const { baseUrl, paths } = tsconfig.compilerOptions
const baseDir = path.resolve(__dirname, baseUrl)

module.exports = ({ platform, dev }, { module, resolve, plugins }) => ({
  entry: `./src/index.ts`,
  module: {
    ...module,
    rules: [
      {
        test: /\.tsx?$/,
        loader: 'ts-loader',
      },
      ...module.rules,
    ],
  },
  resolve: {
    ...resolve,
    extensions: [
      '.ts',
      '.tsx',
      `.${platform}.ts`,
      '.native.ts',
      `.${platform}.tsx`,
      '.native.tsx',
      ...resolve.extensions,
    ],
    alias: {
      assets: path.resolve(__dirname, 'assets'),
    },
    modules: [path.resolve(__dirname, 'src'), 'node_modules'],
  },
  plugins: [...plugins, new Dotenv({ path: dev ? '.env.local' : '.env.prod' })],
})
```

### Typescript configuration (tsconfig.json)
```json
{
  "compilerOptions": {
    "baseUrl": "src",
    "paths": {
      "src/*": ["src/*"],
      "assets/*": ["../assets/*"]
    },
    "target": "es5",
    "module": "esnext",
    "lib": ["es2017", "dom", "esnext.asynciterable"],
    "moduleResolution": "node",
    "jsx": "react",
    "outDir": "artifacts",
    "rootDirs": ["src", "assets"],
    "sourceMap": true,
    "allowJs": false,
    "allowSyntheticDefaultImports": true,
    "forceConsistentCasingInFileNames": true,
    "preserveConstEnums": true,
    "noImplicitAny": false,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noUnusedParameters": true,
    "noUnusedLocals": true,
    "strictNullChecks": true,
    "plugins": [
      {
        "name": "typescript-styled-plugin"
      },
      {
        "name": "typescript-snapshots-plugin",
        "testBlockIdentifiers": ["test"]
      }
    ]
  },
  "filesGlob": ["typings/index.d.ts", "src/**/*.ts", "src/**/*.tsx"],
  "types": ["react", "react-native", "jest"],
  "exclude": ["android", "ios", "build", "node_modules"],
  "compileOnSave": false
}
```

### Your Environment
Windows 10 x64

| software         | version
| ---------------- | -------
| Haul             |  1.0.0-rc.0
| react-native     |  0.55.3
| node             | 9.11.1
| yarn      | 1.5.1
| ts-loader | 4.2.0

**Comment by FredyC** [comment]
Strange. I've updated to 1.0.0-rc.0 just because previously used `1.0.0-beta.13` seemed to be stuck when running `gradlew assembleRelease`. However, I have tried that again now and build is finished successfully and app is working 🎉 

Looks like that error above is some sort of regression, most likely coming from Webpack 4.

**Comment by FredyC** [comment]
It bundles successfully with beta.13, so yea, it's still an issue if I would like to update later. I cannot imagine what other loaders/plugins might be faulty. There is only the `ts-loader` which I updated to latest and `dotenv-webpack` which is just a tiny wrapper around `DefinePlugin` and coming from Webpack itself.

I will try searching through my dependencies for that `getResourceIdentifier` to get some clue.

**Comment by thymikee** [comment]
Just to be sure try to remove node_modules too ;)

**Comment by zamotany** [comment]
This is definitely webpack related problem. From quick search it looks like it might be failing here: 			https://github.com/webpack/webpack/blob/master/lib/Compilation.js#L403

**Comment by jukben** [comment]
Hey! I had the same problem in our client's codebase. 

But it's weird that you have got such a cryptic message https://github.com/callstack/haul/commit/0f713231023fb06cb1cf18b395894d436df461ca should clearly show that it's Webpack issue.

In my case the `dotenv-webpack` was the issue, could you please try to remove it and try it again? I hope that something is wrong with it.

**Comment by FredyC** [comment]
Sorry, I am on a different project for another week, so I don't have a capacity to fiddle with this.

Looks like it's bug in Webpack itself ... https://github.com/webpack/webpack/issues/6675

**Comment by thymikee** [comment]
Thank you @FredyC! Let's keep it to track progress

**Comment by jukben** [comment]
I think I've found the issue! 

In my case it was because the `node_modules/webpack` was version 3.x – required by `storybook`, however, Haul is using version 4. Thus once the `Dotenv` plugin tried to add `webpack.DefinePlugin` it looked up for version 3.x, not 4.x thus the issue raised.

Easiest fix is to add `"webpack": "^4.16.0"` into dependencies, problem is gone. 

Finally.

**Comment by FredyC** [comment]
Nice finding @jukben. I would just add it might be better to use [Yarn resolutions](https://yarnpkg.com/en/docs/selective-version-resolutions) instead of installing dependency you are not using directly.
